### PR TITLE
Fips addons

### PR DESCRIPTION
**Pull request by jschmidb**
This pull request provides additional patches for libica FIPS 140-3 compliance. 
After merging the pr, a release patch will be added to upgrade libica to 4.1.0.

### CODE REVIEW

File: src/s390_ecc.c [hunk 1]
```diff
@@ -2468,55 +2468,124 @@ unsigned int eckeygen_sw(ICA_EC_KEY *key)
...
-	EVP_PKEY *privkey = NULL, *pubkey = NULL;
-	BIGNUM *d = NULL, *x = NULL, *y = NULL;
-	int privlen, rc = EINVAL;
+	EC_KEY *eckey;
+	EC_POINT *point;
+	unsigned char pubbuf[133];
+	size_t publen;
+	int match = 0;
+
+	eckey = EVP_PKEY_get1_EC_KEY(pkey);
```
Comment (ifranzki) [inline src/s390_ecc.c:2486]: `EVP_PKEY_get1_EC_KEY()` is deprectated with OpenSSL 3.0.
`Same for EC_KEY_get0_public_key()`.
Use `EVP_PKEY_todata()` for the OpenSSL 3.0 case.

Comment (jschmidb) [reply src/s390_ecc.c:2486]: OK, after our internal discussion I'm removing the pubkey_match() routine, which also removes the use of deprecated openssl functions.

**Review by ifranzki — APPROVED** [review]
lgtm

File: src/fips.c [hunk 2]
```diff
@@ -95,12 +96,49 @@ static int rsa_kat(void);
...
+	size_t i;
+	int ret = 0;
+
+	if (ica_open_adapter(&ah))
+		return ret;
+
+	if (ica_get_functionlist(NULL, &count) != 0)
+		return ret;
+
+	libica_func_list = malloc(sizeof(libica_func_list_element) *count);
```
Comment (holger-dengler) [inline src/fips.c:113]: Please add a space between * and count.
`malloc(sizeof(libica_func_list_element) * count);`

File: src/fips.c [hunk 3]
```diff
@@ -153,7 +193,7 @@ static inline int sha3_available(void)
 	rc = ica_sha3_224(SHA_MSG_PART_ONLY, sizeof(test_data), test_data,
 			&sha3_224_context, output_hash);
 
-	return (rc == ENODEV ? 0 : 1);
+	return (rc != 0 ? 0 : 1);
```
Comment (holger-dengler) [inline src/fips.c:196]: `return !rc;`?

File: src/ica_api.c [hunk 4]
```diff
@@ -1367,8 +1367,12 @@ int ica_ec_key_init(const unsigned char *X, const unsigned char *Y,
 #ifdef ICA_FIPS
 	if (fips >> 1)
 		return EACCES;
-	if (!curve_supported_via_openssl(key->nid) || !curve_supported_via_cpacf(key->nid))
-		return EPERM;
+	if (fips & ICA_FIPS_MODE) {
```
Comment (holger-dengler) [inline src/ica_api.c:1370]: For all these ICA_FIPS_MODE tests (here and in the following commits); Why not combining them into one if? AFAIK there is only one single path.

File: test/ecdh_test.c [hunk 5]
```diff
@@ -326,6 +326,15 @@ int main(int argc, char **argv)
 
 	/* Iterate over curves */
 	for (i = 0; i < NUM_ECDH_TESTS; i++) {
+
```
Comment (holger-dengler) [inline test/ecdh_test.c:329]: Remove blank line

File: test/ecdsa_test.c [hunk 6]
```diff
@@ -212,6 +212,15 @@ int main(int argc, char **argv)
 
 	/* Iterate over curves */
 	for (i = 0; i < NUM_ECDSA_TESTS; i++) {
+
```
Comment (holger-dengler) [inline test/ecdsa_test.c:215]: Same here

**Review by holger-dengler — APPROVED** [review]
With or without the proposed changes: ACK

File: src/fips.c (hunk 2 above)
Comment (jschmidb) [reply src/fips.c:113]: OK, done.

File: test/ecdh_test.c (hunk 5 above)
Comment (jschmidb) [reply test/ecdh_test.c:329]: You are right, according to coding standards the blank lines are superfluous. I added them intentionally to make the code better readable. I'll check after this pr if I can probably make the loops smaller by putting code into subroutines. There are many places where test routines get lengthy ...

File: src/ica_api.c (hunk 4 above)
Comment (jschmidb) [reply src/ica_api.c:1370]: Yes, the two checks curve_supported_via_openssl() and curve_supported_via_cpacf() are used at several places together. Will check to consolidate this after this pr.

**Pull request merged**